Frameworks/Foundation/NSScanner.mm

```diff
@@ -268,62 +234,93 @@ - (BOOL)scanFloat:(float*)valuep {
  @Status Interoperable
 */
 - (BOOL)scanDouble:(double*)valuep {
-    char* pScanStart = (char*)[_string UTF8String];
-    char* pScanEnd = NULL;
+    NSUInteger startLocation = [self _indexOfNextUnskippedCharacter];
+    const NSUInteger length = _string.length;
+    if (startLocation >= length) {
+        return NO;
+    }
 
-    pScanStart += _location;
+    char* pScan = (char*)[_string UTF8String] + startLocation;
 
-    double val = strtod(pScanStart, &pScanEnd);
-    assert(pScanEnd != NULL);
+    // strtod automatically takes care of spaces (needs to be explicitly specified in charactersToBeSkipped),
+    // and only expects '.' as a decimal separator
+    // copy a scannable range of the string to a new buffer, adjusting for these constraints
+    char decimalSeparator;
+    if ((!_locale) ||
+        (![[_locale objectForKey:NSLocaleDecimalSeparator] getCString:&decimalSeparator maxLength:1 encoding:NSASCIIStringEncoding])) {
+        decimalSeparator = '.';
+    }
 
-    _location += pScanEnd - pScanStart;
+    std::vector<char> charsToScan;
+    charsToScan.reserve(_string.length - startLocation);
+    for (NSUInteger i = startLocation; i < length; ++i, ++pScan) {
+        if ((*pScan >= '0') && (*pScan <= '9')) {
+            charsToScan.push_back(*pScan);
 
-    if (valuep) {
-        *valuep = val;
+        } else if (*pScan == decimalSeparator) {
+            // Add a decimal separator that strtod will accept
+            charsToScan.push_back('.');
+
+        } else {
+            break;
+        }
+    }
+
+    if (charsToScan.size() == 0) {
+        return NO;
+    }
+
+    char* pScanEnd = nullptr;
+    double val = strtod(charsToScan.data(), &pScanEnd);
+
+    if ((!pScanEnd) || (pScanEnd == charsToScan.data())) {
+        return NO;
     }
 
-    if ((pScanEnd - pScanStart) == 0) {
-        return FALSE;
+    _scanLocation = startLocation + (pScanEnd - charsToScan.data());
+
+    if (valuep) {
+        *valuep = val;
     }
 
-    return TRUE;
+    return YES;
 }
 
 /**
  @Status Interoperable
 */
 - (BOOL)scanHexInt:(unsigned*)valuep {
```

Comment:
can we implement scanHexInt in terms of scanHexLongLong? we'd have to accoutn for overflow behaviour, but it would kill half our hex reading code